### DISCUSSION TITLE
setting scalar variable parameter in action

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#25715] [Q&A General] [answered]
I have a NodalKernel that needs a coupled scalar variable. The parameter is set like `params.addRequiredCoupledVar("lambda")` in my `MyNodalKernel`. It works when I have the kernels and variables all written out in my input file. Now I'm trying to make an action.

In my action I create the scalar variable:
```
auto params = _factory.getValidParams("MooseVariable");
params.set<MooseEnum>("order") = "FIRST";
params.set<MooseEnum>("family") = "SCALAR";
_problem->addVariable("MooseVariable", "lambda", params);
```

Then when adding the kernel I do:
```
std::string var = "lambda";
if (_current_task == "add_nodal_kernel") {
  if (_problem->hasScalarVariable(var)) {
    auto params = _factory.getValidParams("MyNodalKernel");
    params.set<std::vector<VariableName>>("lambda") = {var};  // note this wants a vector for some reason
    (... more params...)
    _problem->addNodalKernel("MyNodalKernel", "foo", params);
  }
}
```

When I go to create the object (`_problem->addNodalKernel...`) I get an runtime error: `Scalar variable 'lambda' does not exist in this system`, even though I explicitly tested for it above!

(Also note the comment above where the parameter needs to be set to a vector of variables. If I just pass the variable itself, the code compiles but I get a different runtime error about "VariableName should be std::vector<VariableName>...", but that's a different issue...)

**Comment by jessecarterMOOSE** [comment]
The error about the variable needing to be a vector of variables is:
```
Attempting to set parameter "lambda" with type (VariableName)
but the parameter already exists as type (std::vector<VariableName, std::allocatorVariableName> >)
```
Which I get when I do `params.set<VariableName>("lambda") = var` but I can fix that by passing the variable as a one-item vector as you see above.

**Reply by GiudGiud** [reply]
do this
```
params.set<std::vector<VariableName>>("lambda") = {var};
```

**Reply by jessecarterMOOSE** [reply]
Yes, I did that (though it doesn't make sense). That's not my actual issue though.

**Reply by GiudGiud** [reply]
the coupleable interface works with vectors of variable names.
params.addCoupledVar is roughly the same as
`params.addParam<std::vector<VariableName>>`
with a bit of extra checking

**Reply by GiudGiud** [reply]
I think what you would want here is a `setCoupledScalarVar` routine. We could add it. 
It s just a proliferation of APIs that do roughly the same thing

**Reply by GiudGiud** [reply]
ok getting to your issue now

**Reply by GiudGiud** [reply]
can you do this instead

    getProblem().addVariable("MooseVariableScalar", var_name, params);

**Reply by jessecarterMOOSE** [reply]
yes that did it!

**Reply by jessecarterMOOSE** [reply]
OK, next issue, below that I have a ScalarKernel (actually inherits from NodalScalarKernel), and I set the "variable" parameter like: `params.set<VariableName>("variable") = lambda;`

I get a runtime error:
```
Attempting to set parameter "variable" with type (VariableName)
but the parameter already exists as type (NonlinearVariableName)
```

The error goes away if I change the type to `NonlinearVariableName`, but this is awkward because the variable is a scalar, which I would imagine is distinctively different from a nonlinear variable.

**Reply by GiudGiud** [reply, marked as answer]
scalar variables can be nonlinear.

You ll have to comply with the variable name type haha
